otter/models/cass.py

```diff
@@ -1831,43 +1826,53 @@ def _dict(r): return json.loads(r['server_blob'])
 
         yield do_return((list(rfunc(rows)), last_update))
 
-    def insert_servers(self, last_update, servers, clear_others):
-        """
-        See :method:`IScalingGroupServersCache.insert_servers`
-        """
-        if len(servers) == 0:
-            if clear_others:
-                return self.delete_servers()
-            else:
-                return Effect(Constant(None))
-        query = ('INSERT INTO {cf} ("tenantId", "groupId", last_update, '
-                 'server_id, server_blob, server_as_active) '
-                 'VALUES(:tenantId, :groupId, :last_update, :server_id{i}, '
-                 ':server_blob{i}, :server_as_active{i});')
-        params = merge(self.params, {"last_update": last_update})
-        queries = []
-        for i, server in enumerate(servers):
-            params['server_id{}'.format(i)] = server['id']
-            params['server_as_active{}'.format(i)] = server.pop(
-                '_is_as_active', False)
-            params['server_blob{}'.format(i)] = json.dumps(server)
-            queries.append(query.format(cf=self.table, i=i))
-        if clear_others:
-            return self.delete_servers().on(
-                lambda _: cql_eff(
-                    batch(queries, get_client_ts(self.clock)), params))
-        else:
-            return cql_eff(batch(queries, get_client_ts(self.clock)), params)
+    @do
+    def update_servers(self, time, servers):
+        """
+        See :method:`IScalingGroupServersCache.update_servers`
+
+        This implementation makes two important assumptions which needs to be
+        taken care by caller:
+        - It is not re-entrant for a given object; i.e a call to this must
```

Comment:
Because this function is called serially by only one node guaranteed by convergence partitioner. The same is for `time` comment below. But yeah the leap seconds can cause issue. I'll create ticket to track this.